src/main/java/org/intellij/sdk/codesync/HandleBuffer.java

```diff
@@ -0,0 +1,313 @@
+            ConfigRepoBranch configRepoBranch = configRepo.getRepoBranch(diffFile.branch);
+            if (diffFile.isNewFile) {
+                newFiles.add(diffFile.fileRelativePath);
+                handleNewFile(client, diffFile, configFile, configRepo, configRepoBranch);
+                diffFile.delete();
+            }
+
+            if (newFiles.contains(diffFile.fileRelativePath))  {
+                // Skip the changes diffs if relevant file was uploaded in the same iteration, wait for next iteration
+                continue;
+            }
+
+            if (diffFile.isRename) {
+                if (newFiles.contains(diffFile.oldRelativePath)) {
+                    // If old_rel_path uploaded in the same iteration, wait for next iteration
+                    continue;
+                }
+
+                Integer oldFileId = configRepoBranch.getFileId(diffFile.oldRelativePath);
+                // TODO: old_file_id = config_files.pop(old_rel_path, "") maybe we need to update configFile file here.
+                if (oldFileId == null) {
+                    System.out.printf(
+                        "old_file: %s was not synced for rename of %s/%s",
+                        diffFile.oldRelativePath, diffFile.repoPath, diffFile.fileRelativePath
+                    );
+                    diffFile.delete();
+                }
+
+                handleFileRename(configFile, configRepo, configRepoBranch, diffFile, oldFileId);
+            }
+
+            if (!diffFile.isBinary && !diffFile.isDeleted && diffFile.diff.length() == 0) {
+                System.out.printf("Empty diff found in file: %s. Removing...", diffFile.fileRelativePath);
+                // Delete empty diff files.
+                diffFile.delete();
+                continue;
+            }
+
+            Integer fileId = configRepoBranch.getFileId(diffFile.fileRelativePath);
+            if (fileId == null && !diffFile.isRename &&  !diffFile.isDeleted) {
+                System.out.printf("File ID not found for; %s", diffFile.fileRelativePath);
+                continue;
+            }
+            if (fileId == null && diffFile.isDeleted) {
+                if (!Utils.isDirectoryDelete(diffFile.repoPath, diffFile.branch, diffFile.fileRelativePath)) {
+                    System.out.printf("is_deleted non-synced file found: %s/%s", diffFile.repoPath, diffFile.fileRelativePath);
+                }
+                diffFile.delete();
+
+                continue;
+            }
+
+            if (diffFile.isDeleted){
+                // TODO: See of this can be moved to DiffFile.
+                diffFile.setDiff(
+                    getDiffOfDeletedFile(configFile, configRepo, configRepoBranch, diffFile)
+                );
+            }
+
+            Boolean is_success = webSocketClient.sendDiff(diffFile);
+            if (!is_success) {
+                continue;
+            }
+            diffFile.delete();
+
+            System.out.println("Testings");
```

Comment:
Probably that's on TODO OR I have missed, how are we calling handleBuffer recursivley?

Reply:
It is a TODO. Any suggestions for this?

Reply:
on JS side, I am simply calling same function again in the end after 5s wait